<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        <meta name="author" content="Huwenbo Shi">
        <link rel="canonical" href="https://huwenboshi.github.io/hess/faq/">
        <link rel="shortcut icon" href="../img/favicon.ico">
        <title>FAQ - HESS</title>
        <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="../css/font-awesome-4.5.0.css" rel="stylesheet">
        <link href="../css/base.css" rel="stylesheet">
        <link rel="stylesheet" href="../css/highlight.css">
        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->
	
	<script src="../js/jquery-1.10.2.min.js"></script>
        <script src="../js/bootstrap-3.0.3.min.js"></script>
        <script src="../js/highlight.pack.js"></script>
        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

            ga('create', 'UA-XXXXXXXX-X', 'auto');
            ga('send', 'pageview');
        </script> 
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a class="navbar-brand" href="..">HESS</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                    <li >
                        <a href="..">Home</a>
                    </li>
                    <li >
                        <a href="../input_format/">Input Format</a>
                    </li>
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Estimation <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                            
<li >
    <a href="../local_hsqg/">Local SNP-heritability</a>
</li>
                            
<li >
    <a href="../local_rhog/">Local genetic covariance (correlation)</a>
</li>
                        </ul>
                    </li>
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Analysis <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                            
<li >
    <a href="../viz/">Visualizing local estimates</a>
</li>
                            
<li >
    <a href="../contrast_polygenicity/">Contrast polygenicity</a>
</li>
                            
<li >
    <a href="../infer_causality/">Infer putative causality</a>
</li>
                        </ul>
                    </li>
                    <li class="active">
                        <a href="./">FAQ</a>
                    </li>
                </ul>

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                    <li >
                        <a rel="next" href="../infer_causality/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li class="disabled">
                        <a rel="prev" >
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                    <li>
                        <a href="https://github.com/huwenboshi/hess">
                                <i class="fa fa-github"></i>GitHub
                        </a>
                    </li>
            </ul>
        </div>
    </div>
</div>

        <div class="container">
                <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="main active"><a href="#frequently-asked-questions">Frequently Asked Questions</a></li>
            <li><a href="#what-is-the-sample-size-requirement-for-hess">What is the sample size requirement for HESS?</a></li>
            <li><a href="#why-is-it-necessary-to-re-inflate-genomic-control-factor">Why is it necessary to re-inflate genomic control factor?</a></li>
            <li><a href="#why-do-i-get-negative-local-snp-heritability-estimates">Why do I get negative local SNP-heritability estimates?</a></li>
            <li><a href="#why-do-i-get-nan-for-genome-wide-snp-heritability-variance-estimates">Why do I get NaN for genome-wide SNP-heritability variance estimates?</a></li>
            <li><a href="#why-do-i-get-negative-variance-estimates-for-local-snp-heritability-genetic-covariance">Why do I get negative variance estimates for local SNP-heritability / genetic covariance?</a></li>
            <li><a href="#why-does-hess-give-different-estimate-than-ldsc">Why does HESS give different estimate than LDSC?</a></li>
            <li><a href="#what-does-the-following-error-message-mean">What does the following error message mean?</a></li>
            <li><a href="#what-does-the-following-error-message-mean_1">What does the following error message mean?</a></li>
    </ul>
</div></div>
                <div class="col-md-9" role="main">

<h1 id="frequently-asked-questions">Frequently Asked Questions<a class="headerlink" href="#frequently-asked-questions" title="Permanent link">&para;</a></h1>
<p>This page addresses some frequently asked questions from users.
Most of the question that applies to HESS also applies to \(\rho\)-HESS.</p>
<h2 id="what-is-the-sample-size-requirement-for-hess">What is the sample size requirement for HESS?<a class="headerlink" href="#what-is-the-sample-size-requirement-for-hess" title="Permanent link">&para;</a></h2>
<p>We recommend to apply HESS on GWAS with sample size greater
than 50,000. HESS tends to be downwardly biased at smaller
GWAS since fewer eigenvectors are necessary in the truncated-SVD
regularization to obtain a stable estimate.</p>
<p>On the other hand, if genome-wide SNP-heritability is available for the GWAS
(e.g. estimated from individual-level data), then it may still be possible to
obtain the local estimates
(see <a href="http://huwenboshi.github.io/hess/local_hsqg/#running-the-tool-using-total-snp-heritability">here</a>).</p>
<h2 id="why-is-it-necessary-to-re-inflate-genomic-control-factor">Why is it necessary to re-inflate genomic control factor?<a class="headerlink" href="#why-is-it-necessary-to-re-inflate-genomic-control-factor" title="Permanent link">&para;</a></h2>
<p>Most GWAS studies apply genomic control factor (\(\lambda_{GC}\)) on the
association statistics. This could result in a downward bias in the estimated
heritability since the second step of HESS involves an estimation of the
environmental effect variance (\(\sigma_e^2\)) -- with \(\lambda_{GC}\)
corrected GWAS summary stats, HESS tends to overestimate \(\sigma_e^2\),
resulting in downward bias in local SNP-heritability.</p>
<p>See <a href="http://huwenboshi.github.io/hess/local_hsqg/#note-on-re-inflating-92lambda_gc92">here</a>
for how to specify \(\lambda_{GC}\) in estimating local SNP-heritability,
and <a href="http://huwenboshi.github.io/hess/local_rhog/#note-on-re-inflating-92lambda_gc92_1">here</a>
for genetic covariance.</p>
<h2 id="why-do-i-get-negative-local-snp-heritability-estimates">Why do I get negative local SNP-heritability estimates?<a class="headerlink" href="#why-do-i-get-negative-local-snp-heritability-estimates" title="Permanent link">&para;</a></h2>
<p>Local SNP-heritability are not constrained to be greater than 0, so that if a locus
truly has zero SNP-heritability, then HESS can give a zero estimation in the expectation.</p>
<p>If a locus has negative SNP-heritability estimate, then this is likely because the
local SNP-heritability of the trait is close to zero.</p>
<h2 id="why-do-i-get-nan-for-genome-wide-snp-heritability-variance-estimates">Why do I get NaN for genome-wide SNP-heritability variance estimates?<a class="headerlink" href="#why-do-i-get-nan-for-genome-wide-snp-heritability-variance-estimates" title="Permanent link">&para;</a></h2>
<p>This is because the estimated variance of the genome-wide SNP-heritability
is negative, usually caused by relatively small sample size of the GWAS.</p>
<h2 id="why-do-i-get-negative-variance-estimates-for-local-snp-heritability-genetic-covariance">Why do I get negative variance estimates for local SNP-heritability / genetic covariance?<a class="headerlink" href="#why-do-i-get-negative-variance-estimates-for-local-snp-heritability-genetic-covariance" title="Permanent link">&para;</a></h2>
<p>The variance estimates for local SNP-heritbility (genetic covariance) is a
random variable that is not constrained to be positive. For some loci, the
local SNP-heritability variance estimates may be negative.</p>
<p>Usually, negative variance estimates are caused by relatively small sample
size of the GWAS.</p>
<h2 id="why-does-hess-give-different-estimate-than-ldsc">Why does HESS give different estimate than LDSC?<a class="headerlink" href="#why-does-hess-give-different-estimate-than-ldsc" title="Permanent link">&para;</a></h2>
<p>Overall, HESS and LDSC should give similar estimates. However, HESS and LDSC
have different definitions of SNP-heritability and genetic covariance. Also,
unlike LDSC, HESS does not account for population stratification in the GWAS
summary stats. This could also cause the difference.</p>
<h2 id="what-does-the-following-error-message-mean">What does the following error message mean?<a class="headerlink" href="#what-does-the-following-error-message-mean" title="Permanent link">&para;</a></h2>
<p>Some users experience the following error message in step 2 of the analysis.</p>
<div class="codehilite"><pre>[ERROR] Rank of A less than the number of loci. There might be loci with no SNP.
</pre></div>


<p>The second step of HESS and \(\rho\)-HESS involves inverting a matrix. If
the matrix is not invertible, HESS and \(\rho\)-HESS will not attempt to
estimate SNP-heritability or genetic covariance.</p>
<p>Usually, this error is caused by empty locus (SNP with no SNP in it).</p>
<p>If there are empty loci in the data, it is recommended to remove these loci
or combine these loci with nearby loci.</p>
<h2 id="what-does-the-following-error-message-mean_1">What does the following error message mean?<a class="headerlink" href="#what-does-the-following-error-message-mean_1" title="Permanent link">&para;</a></h2>
<p>Some users experience the following error message in step 2 of the analysis.</p>
<div class="codehilite"><pre>IOError: CRC check failed ...
</pre></div>


<p>This is usually caused by prematurely finished step 1.</p>
<p>Note that step 1 usually involves inverting large matrices and may take quite
bit of memory.</p></div>
        </div>

        <footer class="col-md-12">
            <hr>
                <p>Copyright &copy; 2017 - 2022 Huwenbo Shi</p>
            <p>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>
        <script>var base_url = '..';</script>
        <script data-main="../mkdocs/js/search.js" src="../mkdocs/js/require.js"></script>
        <script src="../js/base.js"></script>
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
        <script src="../assets/mathjaxhelper.js"></script><div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
                <h4 class="modal-title" id="exampleModalLabel">Search</h4>
            </div>
            <div class="modal-body">
                <p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
                <form role="form">
                    <div class="form-group">
                        <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                    </div>
                </form>
                <div id="mkdocs-search-results"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>

    </body>
</html>
