<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        <meta name="author" content="Huwenbo Shi">
        <link rel="canonical" href="https://huwenboshi.github.io/hess/local_hsqg/">
        <link rel="shortcut icon" href="../img/favicon.ico">
        <title>Local SNP-heritability - HESS</title>
        <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="../css/font-awesome-4.5.0.css" rel="stylesheet">
        <link href="../css/base.css" rel="stylesheet">
        <link rel="stylesheet" href="../css/highlight.css">
        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->
	
	<script src="../js/jquery-1.10.2.min.js"></script>
        <script src="../js/bootstrap-3.0.3.min.js"></script>
        <script src="../js/highlight.pack.js"></script>
        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

            ga('create', 'UA-XXXXXXXX-X', 'auto');
            ga('send', 'pageview');
        </script> 
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a class="navbar-brand" href="..">HESS</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                    <li >
                        <a href="..">Home</a>
                    </li>
                    <li >
                        <a href="../input_format/">Input Format</a>
                    </li>
                    <li class="dropdown active">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Estimation <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                            
<li class="active">
    <a href="./">Local SNP-heritability</a>
</li>
                            
<li >
    <a href="../local_rhog/">Local genetic covariance (correlation)</a>
</li>
                        </ul>
                    </li>
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Analysis <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                            
<li >
    <a href="../viz/">Visualizing local estimates</a>
</li>
                            
<li >
    <a href="../contrast_polygenicity/">Contrast polygenicity</a>
</li>
                            
<li >
    <a href="../infer_causality/">Infer putative causality</a>
</li>
                        </ul>
                    </li>
                    <li >
                        <a href="../faq/">FAQ</a>
                    </li>
                </ul>

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                    <li >
                        <a rel="next" href="../input_format/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../local_rhog/">
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                    <li>
                        <a href="https://github.com/huwenboshi/hess">
                                <i class="fa fa-github"></i>GitHub
                        </a>
                    </li>
            </ul>
        </div>
    </div>
</div>

        <div class="container">
                <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="main active"><a href="#estimating-local-snp-heritability">Estimating local SNP-heritability</a></li>
            <li><a href="#step-1-compute-eigenvalues-and-squared-projections">Step 1 - compute eigenvalues and squared projections</a></li>
            <li><a href="#step-2-estimate-local-snp-heritability-and-standard-error">Step 2 - estimate local SNP-heritability and standard error</a></li>
    </ul>
</div></div>
                <div class="col-md-9" role="main">

<h1 id="estimating-local-snp-heritability">Estimating local SNP-heritability<a class="headerlink" href="#estimating-local-snp-heritability" title="Permanent link">&para;</a></h1>
<p>This page describes the steps to estimate local SNP-heritability from GWAS
summary association data. In the first step, HESS computes the eigenvalues of
LD matrices, and the squared projections of GWAS effect size vector onto the
eigenvectors of LD matrices. In the second step, HESS uses the output from
step 1 to obtain estimates of local SNP-heritability and their standard errors.</p>
<h2 id="step-1-compute-eigenvalues-and-squared-projections">Step 1 - compute eigenvalues and squared projections<a class="headerlink" href="#step-1-compute-eigenvalues-and-squared-projections" title="Permanent link">&para;</a></h2>
<h3 id="running-the-tool">Running the tool<a class="headerlink" href="#running-the-tool" title="Permanent link">&para;</a></h3>
<p>In this step, HESS computes the eigenvalues, and the squared projections of
GWAS effect size vector onto the eigenvectors of LD matrices. The following
code provides an example of how to perform this tep.</p>
<div class="codehilite"><pre>for chrom in $(seq 22)
do
    python hess.py \
        --local-hsqg &lt;summary statistics&gt; \
        --chrom $chrom \
        --bfile &lt;reference panel in PLINK format for the specific chromosome&gt; \
        --partition &lt;genome partition file for the specific chromosome&gt; \
        --out step1
done
</pre></div>


<p>In the command above, <code>--local-hsqg</code> tells HESS to estimate local
SNP-heritability, and is used to specify the GWAS summary statistics data;
<code>--chrom</code> is used to specify the chromosome number; <code>--bfile</code> is used to
specify the reference panel for the corresponding chromsome; <code>--partition</code> is
used to specify the genome partition file; <code>--out</code> is used to specify the
prefix of the output file. For input file format, please refer to
<a href="https://huwenboshi.github.io/hess-0.5/input_format/">Input Format</a>.</p>
<div style="background-color:rgba(230, 230, 250, 1.0);">
( <b>Note</b>: The above for loop can be parallelized by chromosome in case a
computer cluster is available. In addition, users can specify the minimum
minor allele frequency (MAF) of the SNPs used for estimation through the
"--min-maf" flag. The default MAF threshold is 0.05. )
</div>

<h3 id="interpreting-the-output">Interpreting the output<a class="headerlink" href="#interpreting-the-output" title="Permanent link">&para;</a></h3>
<p>After executing the command above, 4 files will be created for each
chromosome (i.e. 88 files for all 22 chromosomes in total), taking up ~10MB
of space for the entire genome. The following is an example output obtained
for chromosome 22.</p>
<ul>
<li><strong>step1_chr22.info.gz</strong> - contains the locus information, including start and end
positions, number of SNPs, rank of LD matrices, and sample size</li>
</ul>
<div class="codehilite"><pre>16050408        17674294        371     274     91273
17674295        18296087        419     306     89182
18296088        19912357        947     502     90231
...             ...             ...     ...     ...
</pre></div>


<ul>
<li><strong>step1_chr22.eig.gz</strong> - contains the eigenvalues of LD matrix (up to the rank
of the LD matrix) at each locus, one line per locus</li>
</ul>
<div class="codehilite"><pre>39.31792281  31.23990243  23.81549256  23.47296559  20.45343550  ...
48.73186142  26.95692375  25.32769526  22.11750791  20.55766423  ...
82.58157342  67.42588424  59.52766188  43.10471854  32.15181631  ...
...          ...          ...          ...          ...          ...
</pre></div>


<ul>
<li><strong>step1_chr22.prjsq.gz</strong> - contains the squared projections of GWAS effect size
vector onto the eigenvectors of LD matrix (up to the rank of the LD matrix) at
each locus, one line per locus</li>
</ul>
<div class="codehilite"><pre>0.00008940  0.00001401  0.00013805  0.00009906  0.00007841  ...
0.00054948  0.00001756  0.00008532  0.00002303  0.00004706  ...
0.00008693  0.00005737  0.00070234  0.00008411  0.00004001  ...
...         ...         ...         ...         ...         ...
</pre></div>


<ul>
<li><strong>step1_chr22.log</strong> - contains helpful information for debugging, including
number of SNPs, number of SNPs filtered, etc.</li>
</ul>
<div class="codehilite"><pre>[INFO] Command started at: ...
[INFO] Command issued: ...
[INFO] Loaded ... partitins in ...
[INFO] Average window size is ...
[INFO] ... SNPs read from reference panel
                     ...
[INFO] Command finished at: ...
</pre></div>


<div style="background-color:rgba(230, 230, 250, 1.0);">
( <b>Note</b>: Log files are very useful in solving issues with running the
software. Please include the log file in the email when you report any issue
with HESS. )
</div>

<h2 id="step-2-estimate-local-snp-heritability-and-standard-error">Step 2 - estimate local SNP-heritability and standard error<a class="headerlink" href="#step-2-estimate-local-snp-heritability-and-standard-error" title="Permanent link">&para;</a></h2>
<h3 id="running-the-tool-without-using-total-snp-heritability">Running the tool (without using total SNP-heritability)<a class="headerlink" href="#running-the-tool-without-using-total-snp-heritability" title="Permanent link">&para;</a></h3>
<p>In the second step, HESS uses output from step 1 across all chromosomes
(step1_chr{1..22}.info.gz, step1_chr{1..22}.eig.gz, step1_chr{1..22}.prjsq.gz)
to obtain local SNP-heritability estimates and their standard errors. The
following command automatically looks for output from step 1 across all
chromosomes with the prefix "step1" to obtain local SNP-heritability estimates.</p>
<div class="codehilite"><pre>python hess.py --prefix step1 --out step2
</pre></div>


<div style="background-color:rgba(240, 128, 128, 0.2);">
( <b>Note</b>: This is the recommended way of running HESS. However, if sample
size of the GWAS is relatively small and total SNP-heritability is known, e.g.
estimated from individual-level data, then one may use the script in the
next section to obtain local SNP-heritability estimates. Also note that no
for loop is needed here. )
</div>

<h3 id="note-on-re-inflating-92lambda_gc92">Note on re-inflating \(\lambda_{GC}\)<a class="headerlink" href="#note-on-re-inflating-92lambda_gc92" title="Permanent link">&para;</a></h3>
<p>Most GWAS summary stats are corrected for genomic control factor
\(\lambda_{GC}\). This could result in a downward bias in the estimated
SNP-heritability. If the GWAS summary stats has been corrected for
\(\lambda_{GC}\), it is recommended to use the following code to perform
step 2.</p>
<div class="codehilite"><pre>python hess.py --prefix step1 --reinflate-lambda-gc &lt;lambda gc to reinflate&gt; \
               --out step2
</pre></div>


<h3 id="running-the-tool-using-total-snp-heritability">Running the tool (using total SNP-heritability)<a class="headerlink" href="#running-the-tool-using-total-snp-heritability" title="Permanent link">&para;</a></h3>
<p>If sample size of the GWAS is relatively small, the standard error of total
SNP-heritability estimates of HESS can be large, which could then lead to
large standard error in the local estimates. However, if an estimate of total
SNP-heritability is available, then one may use the following script to
decompose the total SNP-heritability into each locus.</p>
<div class="codehilite"><pre>python hess.py --prefix step1 --tot-hsqg &lt;total SNP-heritability&gt; &lt;SE&gt; --out step2
</pre></div>


<p>It is recommended to use the following code if the GWAS summary stats are
corrected for \(\lambda_{GC}\).</p>
<div class="codehilite"><pre>python hess.py --prefix step1 --reinflate-lambda-gc &lt;lambda gc to reinflate&gt; \
               --tot-hsqg &lt;total SNP-heritability&gt; &lt;SE&gt; --out step2
</pre></div>


<h4 id="other-available-flags">Other available flags<a class="headerlink" href="#other-available-flags" title="Permanent link">&para;</a></h4>
<ul>
<li><code>--max-num-eig</code>: Specifies the number of eigenvectors to use in the
truncated-SVD for inverting the LD matrix, default 50</li>
<li><code>--min-eigval</code>: Specifies the minimum eigenvalue cut off in
truncated-SVD, default 1.0</li>
<li><code>--reinflate-lambda-gc</code>: Specifies the genomic control factor that has been
applied to the summary statistics, default 1.0</li>
<li><code>--gwse-thres</code>: A threshold to cap the standard error of total
SNP-heritability estimate</li>
</ul>
<div style="background-color:rgba(230, 230, 250, 1.0);">
( <b>Note</b>: We recommend to run HESS on GWAS summary statistics that has
not been applied the genomic control factor, which could results in downward
bias. )
</div>

<p>We provide a simple script (<code>misc/estimate_lambdagc.py</code>) to infer the genomic
control factor to reinflate the summary statistics if it is unknown. The
following script should be executed after step 1 completes.</p>
<div class="codehilite"><pre>python misc/estimate_lambdagc.py --prefix step1
</pre></div>


<h3 id="interpreting-the-output_1">Interpreting the output<a class="headerlink" href="#interpreting-the-output_1" title="Permanent link">&para;</a></h3>
<p>After step 2, 2 files will be created. These include</p>
<ul>
<li><strong>step2.txt</strong>: contains local SNP-heritability estimates for all loci across all
chromosomes. The columns are chromosome number, locus start position, locus
end position, number of SNPs in locus, number of eigenvectors used
in truncated-SVD, estimates of local SNP-heritability, estimated variance
of the local SNP-heritability. We also report standard error and test
statistics.</li>
</ul>
<div class="codehilite"><pre>chr start       end         num_snp k   local_h2g   var         se          z           p
1   10583       1892607     1286    50  0.0001293   1.2819e-08  0.00011322  1.142       0.12673
1   1892607     3582736     3045    50  0.0003232   2.7102e-08  0.00016463  1.9632      0.024809
1   3582736     4380811     1622    50  6.7587e-05  1.282e-08   0.00011323  0.59692     0.27528
1   4380811     5913893     3790    50  3.1445e-05  2.6111e-08  0.00016159  0.1946      0.42285
... ...         ...         ...     ... ...         ...         ...         ...         ...
22  46470495    47596318    2444    50  0.00018004  2.0125e-08  0.00014186  1.2691      0.1022
22  47596318    48903703    2997    50  8.389e-05   2.4784e-08  0.00015743  0.53287     0.29706
22  48903703    49824534    3773    50  6.9487e-06  2.5931e-08  0.00016103  0.043151    0.48279
22  49824534    51243298    2789    50  0.00015227  2.4024e-08  0.000155    0.98241     0.16295
</pre></div>


<ul>
<li><strong>step2.log</strong>: contains useful information for debugging</li>
</ul>
<div class="codehilite"><pre>[INFO] Command started at: Tue, 03 Oct 2017 21:01:34
[INFO] Command issued:
    /u/project/pasaniuc/shihuwen/software/hess/hess.py \
        --prefix ... \
        --out ...
[INFO] Loaded results for ... loci from step 1
[INFO] Using 4381458 SNPs with average sample size ...
[INFO] Re-inflate the summary statistics with lambda_gc: ...
[INFO] Total SNP-heritability estimate: ...
[INFO] Command finished at: ...
</pre></div></div>
        </div>

        <footer class="col-md-12">
            <hr>
                <p>Copyright &copy; 2017 - 2022 Huwenbo Shi</p>
            <p>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>
        <script>var base_url = '..';</script>
        <script data-main="../mkdocs/js/search.js" src="../mkdocs/js/require.js"></script>
        <script src="../js/base.js"></script>
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
        <script src="../assets/mathjaxhelper.js"></script><div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
                <h4 class="modal-title" id="exampleModalLabel">Search</h4>
            </div>
            <div class="modal-body">
                <p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
                <form role="form">
                    <div class="form-group">
                        <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                    </div>
                </form>
                <div id="mkdocs-search-results"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>

    </body>
</html>
